core/src/main/scala/chisel3/Mem.scala

```diff
@@ -60,56 +60,57 @@ sealed abstract class MemBase[T <: Data](val t: T, val length: BigInt) extends H
     */
   def apply(x: Int): T = macro SourceInfoTransform.xArg
 
+  def do_apply_impl(idx: UInt, clock: Clock)(implicit sourceInfo: SourceInfo, compileOptions: CompileOptions): T = {
+    if (clock != clockInst) {
+      Builder.warning("The clock used to initialize the memory is different than the one used to initialize the port. This behavior will be deprecated in 3.6.0")
+    }
+    makePort(sourceInfo, idx, MemPortDirection.INFER, clock)
+  }
+
   /** @group SourceInfoTransformMacro */
   def do_apply(idx: BigInt)(implicit sourceInfo: SourceInfo, compileOptions: CompileOptions): T = {
     require(idx >= 0 && idx < length)
-    apply(idx.asUInt)
+    apply(idx.asUInt, Builder.forcedClock)
   }
 
   /** @group SourceInfoTransformMacro */
   def do_apply(idx: Int)(implicit sourceInfo: SourceInfo, compileOptions: CompileOptions): T =
     do_apply(BigInt(idx))(sourceInfo, compileOptions)
 
+  def apply(x: UInt): T = macro SourceInfoTransform.xArg
+
   /** Creates a read/write accessor into the memory with dynamic addressing.
     * See the class documentation of the memory for more detailed information.
     */
-  def apply(x: UInt): T = macro SourceInfoTransform.xArg
+  def apply(x: UInt, y: Clock): T = macro SourceInfoTransform.xyArg
 
   /** @group SourceInfoTransformMacro */
-  def do_apply(idx: UInt)(implicit sourceInfo: SourceInfo, compileOptions: CompileOptions): T =
-    makePort(sourceInfo, idx, MemPortDirection.INFER)
+  def do_apply(idx: UInt, clock: Clock = Builder.forcedClock)(implicit sourceInfo: SourceInfo, compileOptions: CompileOptions): T =
+    do_apply_impl(idx, clock)
```

Comment:
```suggestion
  def do_apply(idx: UInt)(implicit sourceInfo: SourceInfo, compileOptions: CompileOptions): T =
    do_apply_impl(idx, Builder.forcedClock, true)
    
  def do_apply(idx: UInt, clock: Clock)(implicit sourceInfo: SourceInfo, compileOptions: CompileOptions): T =
    do_apply_impl(idx, clock, false)
```

You need different versions of these because you need to suppress the warning when the clock is passed manually.

Same with versions below.